Add ProcessGroupGloo::createDefaultDevice

pietern opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#26166 Add ProcessGroupGloo::createDefaultDevice**

There were 2 variants to create a new device. One to do so based the
name of a network interface, and one to do so based on a hostname or
address. In the latter, if the address was not specified, it would
lookup the local hostname and try to resolve that. If that failed, the
process would crash.

In this default path, we now try to lookup and use the local hostname,
and if that fails we fallback to using the loopback address.

If the local hostname doesn't resolve to an address that we can bind
to, it is very likely that this process won't join other processes
over the network, and that the user is trying to run a local test.

If this assumption is wrong, the user can override the default
interface selection by setting the environment variable
`GLOO_SOCKET_IFNAME` to the name of the external network interface.

I tested this by changing the local hostname to a bogus name and
confirmed that default initialization works as expected.

Closes #26049.

Differential Revision: [D17397898](https://our.internmc.facebook.com/intern/diff/D17397898)

pietern committed:
Add ProcessGroupGloo::createDefaultDevice

There were 2 variants to create a new device. One to do so based the
name of a network interface, and one to do so based on a hostname or
address. In the latter, if the address was not specified, it would
lookup the local hostname and try to resolve that. If that failed, the
process would crash.

In this default path, we now try to lookup and use the local hostname,
and if that fails we fallback to using the loopback address.

If the local hostname doesn't resolve to an address that we can bind
to, it is very likely that this process won't join other processes
over the network, and that the user is trying to run a local test.

If this assumption is wrong, the user can override the default
interface selection by setting the environment variable
`GLOO_SOCKET_IFNAME` to the name of the external network interface.

I tested this by changing the local hostname to a bogus name and
confirmed that default initialization works as expected.

Closes #26049.

pietern committed:
Update on "Add ProcessGroupGloo::createDefaultDevice"


There were 2 variants to create a new device. One to do so based the
name of a network interface, and one to do so based on a hostname or
address. In the latter, if the address was not specified, it would
lookup the local hostname and try to resolve that. If that failed, the
process would crash.

In this default path, we now try to lookup and use the local hostname,
and if that fails we fallback to using the loopback address.

If the local hostname doesn't resolve to an address that we can bind
to, it is very likely that this process won't join other processes
over the network, and that the user is trying to run a local test.

If this assumption is wrong, the user can override the default
interface selection by setting the environment variable
`GLOO_SOCKET_IFNAME` to the name of the external network interface.

I tested this by changing the local hostname to a bogus name and
confirmed that default initialization works as expected.

Closes #26049.

mrshenli commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -300,6 +303,43 @@ void ProcessGroupGloo::RecvWork::wait() {
+    }
+    rv = bind(fd, rp->ai_addr, rp->ai_addrlen);
+    if (rv == -1) {
+      close(fd);
```
nit: as this is closed anyway, maybe move this to the front of `if(rv == -1)`?

pietern replied:
Good catch. I copied this code from Gloo and removed the parts that were not needed.

mrshenli commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -322,46 +362,78 @@ std::shared_ptr<::gloo::transport::Device> ProcessGroupGloo::
+      "Cannot resolve ",
+      hostname,
+      " to a (local) address");
+  return ::gloo::transport::tcp::CreateDevice(attr);
```
what does `CreateDevice` do in tcp and uv if the hostname is not valid?

pietern replied:
It crashes saying it cannot resolve the hostname or bind to the address. 

This is also what I want to prevent with this commit. Instead, we can either fall back to using the loopback interface when using the default constructor, or raise a runtime error with TORCH_CHECK prior to calling `CreateDevice` when using the non-default constructor.

mrshenli commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -322,46 +362,78 @@ std::shared_ptr<::gloo::transport::Device> ProcessGroupGloo::
+  }
+
+  // Otherwise, use the loopback interface.
+  return createDeviceForInterface("lo");
```
will there be any log telling users we are binding to the loopback or resolved hostname address?

pietern replied:
There isn't now, but there should be one indeed.

pietern committed:
Update on "Add ProcessGroupGloo::createDefaultDevice"


There were 2 variants to create a new device. One to do so based the
name of a network interface, and one to do so based on a hostname or
address. In the latter, if the address was not specified, it would
lookup the local hostname and try to resolve that. If that failed, the
process would crash.

In this default path, we now try to lookup and use the local hostname,
and if that fails we fallback to using the loopback address.

If the local hostname doesn't resolve to an address that we can bind
to, it is very likely that this process won't join other processes
over the network, and that the user is trying to run a local test.

If this assumption is wrong, the user can override the default
interface selection by setting the environment variable
`GLOO_SOCKET_IFNAME` to the name of the external network interface.

I tested this by changing the local hostname to a bogus name and
confirmed that default initialization works as expected.

Closes #26049.

pietern commented:
@mrshenli The output if the hostname of the machine doesn't resolve is:

```
../torch/lib/c10d/ProcessGroupGloo.cpp:444: UserWarning: Unable to resolve hostname to a (local) address. Using the loopback address as fallback. Manually set the network interface to bind to with GLOO_SOCKET_IFNAME.
```
